### PR TITLE
WIP: Convert Outer Join To Inner Join

### PR DESCRIPTION
**Pull request by Pennnnnng**
**What I want to do:**

A LEFT or RIGHT OUTER JOIN is converted to an INNER JOIN if the following condition is true:

1. a null-intolerant predicate referencing columns of the null-supplying tables is present in the query WHERE clause. Since this predicate is null-intolerant, any all-NULL row that would be produced by the OUTER join is eliminated from the result, making the query semantically equivalent to an inner join.

In particular, if the LEFT/RIGHT table of the FULL OUTER  JOIN has a predicate that can filter null, the FULL OUTER  JOIN will be converted to LEFT/RIGHT JOIN.

Benefits of the above optimization:

1. More predicate could be pushdown
2. Better reorder

refer: https://infocenter.sybase.com/help/index.jsp?topic=/com.sybase.help.sqlanywhere.11.0.1/dbusage_en11/queryopt-sectb-5356466.html

**Optimized Case:** 

EXPLAIN SYNTAX
SELECT *
FROM t
LEFT JOIN t1 ON t.a = t1.c
WHERE t1.c > 0

──explain────────────┐
│ SELECT             │
│     a,             │
│     b,             │
│     c,             │
│     d              │
│ FROM t             │
│ ALL INNER JOIN     │
│ (                  │
│     SELECT         │
│         c,         │
│         d          │
│     FROM t1        │
│     PREWHERE c > 0 │
│ ) AS t1 ON a = c   │
│ WHERE c > 0        │
──────────────────┘

**Design:**

The join of ck will been parsed into a left deep tree. If the current node is LEFT/FULL OUTER JOIN, and the right subtree has a null-intolerant predicate, it can be converted to INNER/LEFT JOIN. If the current node is RIGHT/FULL OUTER JOIN, and the left subtree has a null-intolerant predicate, it can be converted to INNER/RIGHT JOIN. ([src/Interpreters/OuterToInnerJoinVisitor.cpp:261 - 286]).

A rough definition of a null-intolerant predicate is as follows:
1. col != NULL
2. col = Non-Null value
3. col > NULL or NULL < col

**Route Map**
- [x] Complete the core code
- [x] Add annotations and necessary anomaly detection
- [ ] Complete testing and performance testing
- [ ] Future work: perfect judgment about null-intolerant predicate

### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
 Convert Outer Join To Inner Join

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

### CODE REVIEW

**Comment by Pennnnnng** [comment]
soryy, I fortge to add Changelog category. May I need colse it and push agagin?

**Comment by robot-clickhouse-ci-2** [comment]
This is an automated comment for commit 53cfd9384454f5482f86a43eca82d2578516d0f9 with description of existing statuses. It's updated for the latest CI running
The full report is available [here](https://s3.amazonaws.com/clickhouse-test-reports/52876/53cfd9384454f5482f86a43eca82d2578516d0f9/ci_running.html)
The overall status of the commit is 🔴 failure

<table><thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody><tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td>🔴 failure</td></tr>
</table>

**Comment by hanfei1991** [comment]
@Pennnnnng please remind me when this pr is ready to review. I'm happy to see this optimization. Thanks!

File: src/Interpreters/OuterToInnerJoinVisitor.h [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+#include <Interpreters/InDepthNodeVisitor.h>
+#include <Interpreters/Aliases.h>
+
+namespace DB
+{
+
+class ASTSelectQuery;
+struct TableWithColumnNamesAndTypes;
+
+/// AST transformer. It replaces outer joins with equivalented inner join if possible.
```
Comment (vdimir) [inline src/Interpreters/OuterToInnerJoinVisitor.h:12]: Could you please add more details which transformations are being made? Tests also will be appreciated.

Comment (Pennnnnng) [reply src/Interpreters/OuterToInnerJoinVisitor.h:12]: Relevant comments have been added for the conversion rule

**Comment by hanfei1991** [comment]
@Pennnnnng After some discussion with @KochetovNicolai , we would prefer to implement it in plan level rather than syntax level. It will be involved with lots of refactor and we plan to implement it step by step.

Thank you for your contribution and we look forward for your further engagement.

**Pull request closed without merging**